### PR TITLE
Add Coroutine Support

### PR DESCRIPTION
**Pull request by xa17d**
Since we need to test coroutines, I added initial coroutine support to sweetest.
This is an initial draft how coroutine testing could look like. Feedback appreciated.

**Overview**
`Steps` now implement also `CoroutineScope`, which means you can use coroutine builders like `launch` and `async` in the steps.

There is now the extension function `testCoroutine` for `BaseJUnitTest`, that allows you to call suspending functions, but still remains compatibility to JUnit. Example:
```
fun `awaitTrigger completes when triggered`() = testCoroutine {
   ...
```

If you are using suspending functions in steps, you should import `invoke` from the `coroutine` package:
`com.mysugr.sweetest.framework.coroutine.invoke`
instead of
`com.mysugr.sweetest.framework.base.invoke`.
This allows to use suspending functions inside the run lambda.

Use `assertSequence` to assert that your code runs in the intended sequence. Example:
```
// this works
assertSequence {
    expect(1)
    expect(2)
    expect(3)
}

// this will fail
assertSequence {
    expect(1)
    expect(3)
    expect(2)
}
```
See `TriggerAwaiterSteps` for a more sophisticated example.

In a test, everything will run on the same thread. Child jobs are spawned with a `SupervisorJob` parent. So an error in a child job won't cancel the whole test.

See the example:
`TriggerAwaiter`
`TriggerAwaiterTest`
`TriggerAwaiterSteps`

### CODE REVIEW

File: app/src/main/java/com/mysugr/android/testing/example/coroutine/TriggerAwaiter.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+class TriggerAwaiter {
+    private var currentContinuation: Continuation<Unit>? = null
+
+    fun trigger() {
+        currentContinuation?.resume(Unit)
+    }
+
+    suspend fun awaitTrigger() = suspendCancellableCoroutine<Unit> { continuation ->
+        if (currentContinuation != null) {
+            throw IllegalStateException("Sorry, only one awaitTrigger at a time")
```
Comment (eder7) [inline app/src/main/java/com/mysugr/android/testing/example/coroutine/TriggerAwaiter.kt]: can be abbreviated with kotlin's `error` function

File: sweetest/src/main/java/com/mysugr/sweetest/framework/coroutine/AssertSequence.kt [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package com.mysugr.sweetest.framework.coroutine
+
+suspend fun assertSequence(block: suspend AssertSequence.() -> Unit) {
+    block(AssertSequence())
+}
+
+class AssertSequence {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/coroutine/AssertSequence.kt]: after some thinking i got behind what's the purpose here, but maybe you could add a line of explanation as kdoc?

File: sweetest/src/main/java/com/mysugr/sweetest/framework/coroutine/CoroutinesTestContext.kt [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import kotlinx.coroutines.newSingleThreadContext
+import kotlin.coroutines.CoroutineContext
+
+class CoroutinesTestContext {
+    private val name = CoroutineName("testCoroutine${instanceCounter++}")
+    private val supervisorJob = SupervisorJob()
+    val coroutineContext: CoroutineContext
+        get() = coroutineDispatcher + supervisorJob + name
+
+    suspend fun testFinished() {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/coroutine/CoroutinesTestContext.kt]: i think it's better to use a name that actually says what the function does (especially if it's something that simple) and/or use verbs for function names, e.g. `cancel`, but i'm not sure if it should also have an explicit `Blocking` suffix?

File: sweetest/src/main/java/com/mysugr/sweetest/framework/coroutine/TestExtensions.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import com.mysugr.sweetest.framework.base.BaseJUnitTest
+import com.mysugr.sweetest.framework.base.Steps
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.withContext
+
+fun BaseJUnitTest.testCoroutine(
+        testBlock: suspend CoroutineScope.() -> Unit
+) {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/coroutine/TestExtensions.kt]: wrong indent or line break?

File: sweetest/src/main/java/com/mysugr/sweetest/framework/coroutine/AssertSequence.kt [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    private var currentSequenceIndex = 0
+    fun expect(sequenceIndex: Int) {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/coroutine/AssertSequence.kt]: it doesn't have the same signature as [kotlin's `expect`](https://kotlinlang.org/api/latest/kotlin.test/kotlin.test/expect.html) but it confused me for a second. i don't know what is better here: simplicity (`expect`) or more specificity (`expectSequence`)?

File: sweetest/src/main/java/com/mysugr/sweetest/framework/coroutine/AssertSequence.kt [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+        currentSequenceIndex++
+        if (sequenceIndex != currentSequenceIndex) {
+            throw AssertionError("Expected $currentSequenceIndex, but reached $sequenceIndex")
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/coroutine/AssertSequence.kt]: could use kotlin's [`fail` function](https://kotlinlang.org/api/latest/kotlin.test/kotlin.test/-default-asserter/fail.html) here

**Comment by eder7** [comment]
@Lois9Y @philipphofmann would really love to have some more feedback or you coroutines specialists 😉

File: sweetest/src/main/java/com/mysugr/sweetest/framework/coroutine/TestExtensions.kt [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    runBlocking {
+        val coroutinesTestContext = accessor.testContext.coroutines
+        withContext(coroutinesTestContext.coroutineContext) {
+            testBlock()
+        }
+        coroutinesTestContext.testFinished()
+    }
+}
+
+suspend operator fun <T : Steps> T.invoke(run: suspend T.() -> Unit) = run(this)
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/coroutine/TestExtensions.kt]: i fear, the use of the same operator `invoke` used in the `base` package will cause to confusion some time. also the `base` package will be frequently `*`-imported. i don't know how to handle this best...

File: sweetest/src/main/java/com/mysugr/sweetest/framework/coroutine/CoroutinesTestContext.kt [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+        supervisorJob.cancelAndJoin()
+    }
+
+    companion object {
+        val coroutineDispatcher by lazy { newSingleThreadContext("testDispatcher") }
```
Comment (Lois9Y) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/coroutine/CoroutinesTestContext.kt]: newSingleThreadContext will be removed in the future:
https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/new-single-thread-context.html 
In order for this not to break based on changes in the by the Kotlin team i suggest something like: 
`val coroutineDispatcher by lazy { Executors.newSingleThreadExecutor().asCoroutineDispatcher() }`

**Pull request merged**